compiler/fm/src/lib.rs

```diff
@@ -37,48 +33,24 @@ impl std::fmt::Debug for FileManager {
 }
 
 impl FileManager {
-    pub fn new(root: &Path, file_reader: Box<FileReader>) -> Self {
+    pub fn new(root: &Path) -> Self {
         Self {
             root: root.normalize(),
             file_map: Default::default(),
             id_to_path: Default::default(),
             path_to_id: Default::default(),
-            file_reader,
         }
     }
 
     pub fn as_file_map(&self) -> &FileMap {
         &self.file_map
     }
 
-    pub fn add_file(&mut self, file_name: &Path) -> Option<FileId> {
-        // Handle both relative file paths and std/lib virtual paths.
-        let resolved_path: PathBuf = if is_stdlib_asset(file_name) {
-            // Special case for stdlib where we want to read specifically the `std/` relative path
-            // TODO: The stdlib path should probably be an absolute path rooted in something people would never create
-            file_name.to_path_buf()
-        } else {
-            self.root.join(file_name).normalize()
-        };
-
-        // Check that the resolved path already exists in the file map, if it is, we return it.
-        if let Some(file_id) = self.path_to_id.get(&resolved_path) {
-            return Some(*file_id);
-        }
-
-        // Otherwise we add the file
-        let source = file_reader::read_file_to_string(&resolved_path, &self.file_reader).ok()?;
-        let file_id = self.file_map.add_file(resolved_path.clone().into(), source);
-        self.register_path(file_id, resolved_path);
-        Some(file_id)
-    }
-
-    // TODO: this will become the default strategy for adding files. Possibly with file_reader.
-    // TODO: we are still migrating to this strategy, so we keep the old one for now.
-    // TODO: For the stdlib crate, we need to do this preemptively due to the way we special handle it
+    // Adds a file to the file manager with its source code
     pub fn add_file_with_source(&mut self, file_name: &Path, source: String) -> Option<FileId> {
+        let file_name = self.root.join(file_name).normalize();
```

Comment:
This is a bit problematic and would make #3760 fail without, in particular `path_resolve_modules_with_different_paths_as_same_file` would fail. 

I need to do a bit more thinking on whether we should be normalizing paths here, given that the same caller should not be giving multiple variations of the same path

Reply:
We could disallow paths like this: `./../../lib.nr` and just say that the user needs to normalize the path with the root joined to it.

As one can see from the test, we are adjoining the root, but we don't want to do this for the stdlib. Ideally we don't have special logic in here for the stdlib

Reply:
Another option is to have:

```
    pub fn add_file_with_source(&mut self, file_name: &Path, source: String) -> Option<FileId> {
        let file_name = self.root.join(file_name).normalize();
        self.add_file_with_source_canonical_path(&file_name, source)
    }
        // Adds a file to the file manager with its source code
    pub fn add_file_with_source_canonical_path(&mut self, file_name: &Path, source: String) -> Option<FileId> {
        let file_name = file_name.normalize();
        // Check that the file name already exists in the file map, if it is, we return it.
        if let Some(file_id) = self.path_to_id.get(&file_name) {
            return Some(*file_id);
        }
        let file_name_path_buf = file_name.to_path_buf();

        // Otherwise we add the file
        let file_id = self.file_map.add_file(file_name_path_buf.clone().into(), source);
        self.register_path(file_id, file_name_path_buf);
        Some(file_id)
    }
    ```
    
    Where stdlib calls `add_file_with_source_canonical_path` and all other crates call `add_file_with_source` -- we may need to modify some of our tests -- possibly not as `root` is usually set to `/`

Reply:
Went with the above strategy, so that we did not need to modify tests and so that FileManager does not special case the stdlib